components/patient-access-rules/rest/src/main/java/org/phenotips/data/permissions/rest/internal/DefaultCollaboratorsResourceImpl.java

```diff
@@ -0,0 +1,268 @@
+            if (StringUtils.isNotBlank(id) && StringUtils.isNotBlank(level)) {
+                return postCollaborator(id, level, patientId);
+            }
+        }
+        this.logger.error("The id, permissions level, or both were not provided or are invalid");
+        throw new WebApplicationException(Response.Status.BAD_REQUEST);
+    }
+
+    @Override
+    public Response deleteCollaborators(String patientId)
+    {
+        return this.updateCollaborators(new LinkedList<Collaborator>(), patientId);
+    }
+
+    @Override
+    public Response putCollaborators(String json, String patientId)
+    {
+        List<Collaborator> collaborators = this.jsonToCollaborators(json);
+        return this.updateCollaborators(collaborators, patientId);
+    }
+
+    private Response postCollaborator(String collaboratorId, String accessLevelName, String patientId)
+    {
+        this.checkCollaboratorInfo(collaboratorId, accessLevelName);
+
+        this.logger.debug(
+            "Adding collaborator [{}] with permission level [{}] to the patient record [{}] via REST",
+            collaboratorId, accessLevelName, patientId);
+        // besides getting the patient, checks that the user has manage access
+        PatientAccessContext patientAccessContext = this.secureContextFactory.getContext(patientId, MANAGE_LEVEL);
+        PatientAccess patientAccess = patientAccessContext.getPatientAccess();
+
+        // will throw an error if something goes wrong
+        this.addCollaborator(collaboratorId, accessLevelName.trim(), patientAccess);
+        return Response.noContent().build();
+    }
+
+    private Response updateCollaborators(Collection<Collaborator> collaborators, String patientId)
+    {
+        // besides getting the patient, checks that the user has manage access
+        PatientAccessContext patientAccessContext = this.secureContextFactory.getContext(patientId, MANAGE_LEVEL);
+        PatientAccess patientAccess = patientAccessContext.getPatientAccess();
+
+        if (!patientAccess.updateCollaborators(collaborators)) {
+            this.logger.error("Could not update collaborators");
+            throw new WebApplicationException(Response.Status.INTERNAL_SERVER_ERROR);
+        }
+        return Response.noContent().build();
+    }
+
+    private void addCollaborator(String id, String levelName, PatientAccess patientAccess)
+        throws WebApplicationException
+    {
+        // checking that the access level is valid
+        AccessLevel level = this.getAccessLevelFromString(levelName);
+        EntityReference collaboratorReference = this.currentResolver.resolve(id, EntityType.DOCUMENT, XWIKI_SPACE);
+
+        // todo. function .addCollaborator has to check if the collaborator already exists before adding them
+        if (!patientAccess.addCollaborator(collaboratorReference, level)) {
+            // todo. should this status be an internal server error, or a bad request?
```

Comment:
Can be because they are unauthorized (but very unlikely). Most likely cause would be DB/XWiki error.